notation.go

```diff
@@ -63,31 +44,189 @@ type SignOptions struct {
+	PluginConfig map[string]string
+}
+
+// ValidationResult encapsulates the verification result (passed or failed)
+// for a verification type, including the desired verification action as
+// specified in the trust policy
+type ValidationResult struct {
+	// Type of verification that is performed
+	Type trustpolicy.ValidationType
+
+	// Action is the intended action for the given verification type as defined
+	// in the trust policy
+	Action trustpolicy.ValidationAction
+
+	// Error is set if there are any errors during the verification process
+	Error error
 }
 
-// Validate does basic validation on VerifyOptions.
-func (opts VerifyOptions) Validate() error {
-	return nil
+// VerificationOutcome encapsulates a signature blob's descriptor, its content,
+// the verification level and results for each verification type that was
+// performed.
+type VerificationOutcome struct {
+	// RawSignature is the signature envelope blob
+	RawSignature []byte
+
+	// EnvelopeContent contains the details of the digital signature and
+	// associated metadata
+	EnvelopeContent *signature.EnvelopeContent
+
+	// VerificationLevel describes what verification level was used for
+	// performing signature verification
+	VerificationLevel *trustpolicy.VerificationLevel
+
+	// VerificationResults contains the verifications performed on the signature
+	// and their results
+	VerificationResults []*ValidationResult
+
+	// Error that caused the verification to fail (if it fails)
+	Error error
 }
 
 // Verifier is a generic interface for verifying an artifact.
 type Verifier interface {
-	// Verify verifies the signature and returns the verified descriptor and
-	// metadata of the signed artifact.
-	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (Descriptor, error)
+	// Verify verifies the signature blob and returns the artifact
+	// descriptor upon successful verification.
+	// If nil signature is present and the verification level is not 'skip',
+	// an error will be returned.
+	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (ocispec.Descriptor, *VerificationOutcome, error)
+}
+
+// Verify performs signature verification on each of the notation supported
+// verification types (like integrity, authenticity, etc.) and return the
+// verification outcomes.
+// For more details on signature verification, see
+// https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#signature-verification
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
+	var verificationOutcomes []*VerificationOutcome
+	artifactRef := opts.ArtifactReference
+	// passing nil signature to check 'skip'
+	_, outcome, err := verifier.Verify(ctx, nil, opts)
+	if err != nil {
+		if outcome == nil {
+			return ocispec.Descriptor{}, nil, err
+		}
+	} else if outcome.VerificationLevel.Name == trustpolicy.LevelSkip.Name {
```

Comment:
Should we compare the level directly instead of comparing the name?
```suggestion
	} else if outcome.VerificationLevel == trustpolicy.LevelSkip {
```

Reply:
This might not work as both of them are pointers to structs? I had to use reflect.DeepEqual to make it work.